File: web-app/src/app/memes/Memes.kt
package app.memes

/**
 * @author sokomishalov
 */
import api.loadMemes
import app.memes.container.memeContainer
import dto.Meme
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import react.*
import react.dom.div

class Memes : RComponent<RProps, Memes.State>() {

    class State(
            var page: Int,
            var memes: List<Meme>
    ) : RState

    override fun State.init() {
        page = 0
        memes = emptyList()
    }

    override fun componentDidMount() {
        GlobalScope.launch { loadMore() }
    }

    private suspend fun loadMore() {
        val newMemes = loadMemes(state.page)
        setState {
            page += 1
            memes += newMemes
        }
    }

    override fun RBuilder.render() {
        div("memes") {
            state.memes.map { memeContainer(it) }
        }
    }
}

fun RBuilder.memes() = child(Memes::class) {
}

File: web-app/src/app/memes/Memes.js
import React, {useState} from 'react'
import "./Memes.css"
import {getMemesPage} from "../../api/memes"
import MemeContainer from "./container/MemeContainer"
import _ from "lodash"
import InfiniteScroll from 'react-infinite-scroller'
import {BackTop} from "antd"
import Loader from "../common/loader/loader"

const Memes = () => {
    const [loading, setLoading] = useState(false)
    const [memes, setMemes] = useState([])
    const [hasMore, setHasMore] = useState(true)

    const loadMore = async (page) => {
        setLoading(true)
        const newMemes = await getMemesPage(page - 1)
        if (!_.isEmpty(newMemes)) {
            setMemes(_.concat(memes, newMemes))
        } else {
            setHasMore(false)
        }
        setLoading(false)
    }

    return (
        <div className="memes">
            <InfiniteScroll pageStart={0}
                            loadMore={loadMore}
                            hasMore={hasMore}
                            loader={<Loader loading={loading}/>}>
                {_.map(memes, (meme) => <MemeContainer key={meme["id"]} meme={meme}/>)}
            </InfiniteScroll>
            <BackTop className="memes-backtop"/>
        </div>
    )
}

export default Memes